_posts/2022-01-18-four-favorite-firmware-debug-tools.md

```diff
@@ -0,0 +1,128 @@
+
+The Saleae device itself is, quite literally, a black box. (Well, unless you get a pink one, like I did.) UI and control happens, instead, through a delightful, easy to use GUI. Logic, Saleae's desktop software, makes it a breeze to navigate logic captures, apply protocol analyzers, and find regions of interest. Scrolling, zooming, and measuring time between edges is delightful and simple. Surprisingly, this is still true with a touch screen laptop. It was a joy to use a Saleae with my touchscreen-enabled daily driver from a few jobs ago. A big rewrite of Saleae's firmware and desktop application over the last few years has made the system real-time capable, too!
+
+Salea also has handy scripting utilities baked into the device. [A Socket API with Python and C# bindings](https://support.saleae.com/saleae-api-and-sdk/socket-api) allows you call all sorts of functionality from your favorite proramming environment. The Python wrapper even comes with [handy command line functionality.](https://github.com/saleae/python-saleae-cli)
+
+### Illogical
+The Saleae is a great tool, but it isn't perfect. It's only tolerant of low voltage IO signals. Anything between 1.2V and 5V, it can handle with no problems. Use it on higher voltage signals, and you're gonna have a bad time. It's also limited to doing relatively low bandwidth digital signals - it caps out at 500MS/s. This is more than adequate for most routine firmware protocols, but for higher speed signals like RGMII, USB3.0, or PCIe, you're going to need a more specialized piece of equipment. 
+
+Additionally - being a logic analyzer, you need to take care that you're not accidentally degrading the signals you're monitoring. This can be particularly tricky on faster signal buses, where edge rates of signals are shrinking into the single digit nanoseconds. (This author was recently re-educated on this matter by a quad SPI bus running at 30MHz.) All is not lost, though! There are a few relatively simple things you can do to eke out all the signal bandwidth you can out of your Saleae:
+
+* Be sure to include the paired ground test lead when connecting the Saleae to your target board. 
+* Twist the ground connection together with the signal wire. That way, you're keeping the impedance of the Saleae's test leads as low as possible.
+* Get rid of the grabber clip if you can. Solder a length of 0.1" header to your target net, and plug the test lead directly to the header.
+* Even better: ask your board designer to drop a set of 0.1" headers onto the board. SMT is preferable, if you have the space. An SMT header causes less interruption of the ground plane, and reduces the chance of causing signal integrity problems due to your debug probe connection. 
+
```

Comment:
Totally up to you, but it might be nice to note that while the excellent Saleae software is not open-source, the devices are supported by the open-source Sigrok tool: https://support.saleae.com/faq/technical-faq/saleae-open-source-support

Reply:
I'm going to pass on talking about Sigrok - I don't know the first thing about it, and I'd rather get you this article on time than delay to do the research being able to speak intelligently about it. (If I did, you might be waiting for a few weeks!)